src/extra/gpkg/read.jl

```diff
@@ -0,0 +1,439 @@
+  end
+end
+
+function parse_wkb_point(io::IOBuffer, byte_order::UInt8, has_z::Bool, has_m::Bool, crs_type)
+  coords_list = read_coordinate_values(io, byte_order, has_z, has_m)
+  return create_point_from_coords(coords_list, crs_type)
+end
+
+function parse_wkb_linestring(io::IOBuffer, byte_order::UInt8, has_z::Bool, has_m::Bool, crs_type; is_ring::Bool=false)
+  read_value = byte_order == 0x01 ? ltoh : ntoh
+  
+  num_points = read_value(read(io, UInt32))
+  
+  # Parse all points using map-do syntax
+  points = map(1:num_points) do i
+    coords_list = read_coordinate_values(io, byte_order, has_z, has_m)
+    create_point_from_coords(coords_list, crs_type)
+  end
+  
+  # Check if WKB writer included duplicate closing point (varies by implementation)
+  has_equal_ends = length(points) >= 2 && points[1] == points[end]
+  
+  # Two-flag logic explained:
+  # - is_ring: Semantic flag indicating this linestring is part of a polygon structure
+  # - has_equal_ends: WKB format flag indicating if writer included duplicate closing point
+  # We need both because polygon rings should always return Ring regardless of WKB format variation
+  if is_ring || has_equal_ends
+    # Return a Ring, exclude duplicate last point if present
+    if has_equal_ends
+      return Ring(points[1:end-1]...)
+    else
+      return Ring(points...)
+    end
+  else
+    # It's an open linestring, return a Rope
+    return Rope(points...)
+  end
+end
+
+function parse_wkb_polygon(io::IOBuffer, byte_order::UInt8, has_z::Bool, has_m::Bool, crs_type)
+  read_value = byte_order == 0x01 ? ltoh : ntoh
+  
+  num_rings = read_value(read(io, UInt32))
+  
+  # Parse exterior ring using ring parsing function
+  exterior_ring = parse_wkb_linestring(io, byte_order, has_z, has_m, crs_type; is_ring=true)
+  
+  # Parse interior rings (holes) using map-do syntax
+  hole_rings = map(2:num_rings) do ring_idx
+    parse_wkb_linestring(io, byte_order, has_z, has_m, crs_type; is_ring=true)
+  end
+  
+  # Always use the second branch approach as suggested in review
+  return PolyArea(exterior_ring, hole_rings...)
+end
+
+
+function parse_wkb_multipoint(io::IOBuffer, byte_order::UInt8, has_z::Bool, has_m::Bool, crs_type)
+  read_value = byte_order == 0x01 ? ltoh : ntoh
+  num_points = read_value(read(io, UInt32))
+  
+  # Parse all points using map-do syntax
+  points = map(1:num_points) do i
+    # Each point has its own header in multi-geometry
+    byte_order_inner = read(io, UInt8)
+    read_value_inner = byte_order_inner == 0x01 ? ltoh : ntoh
+    wkb_type_inner = read_value_inner(read(io, UInt32))
+    
+    # Parse the individual geometry's Z/M flags from its own header
+    inner_base_type = wkb_type_inner & 0x0FFFFFFF
+    inner_has_z = (wkb_type_inner & WKB_Z) != 0
+    inner_has_m = (wkb_type_inner & WKB_M) != 0
+    
+    if inner_base_type != WKB_POINT
+      error("Expected WKB_POINT ($WKB_POINT) in multipoint, got $inner_base_type")
+    end
```

Comment:
I believe this is not possible. If the function for multipoint is being called with an inner geometry that is not a point, that is an issue with the logic of the implementation.